### PR TITLE
Improved MS4525 I2C probing

### PR DESCRIPTION
**Pull request by JulioCesarMatias**
AirSpeed:

- Adds probing for all 3 possible MS4525 I2C addresses.

### CODE REVIEW

**Review by DzikuVx — CHANGES_REQUESTED** [review]
@JulioCesarMatias I'm not sure this is a correct way to do it. I mean, this approach makes initialization even longer and is very low level. What would make more sense is to make I2C address user configurable. So we have a CLI setting to choose in dec: 40, 54, 70

File: src/main/drivers/pitotmeter/pitotmeter_ms4525.c [hunk 1]
```diff
@@ -113,7 +114,8 @@ bool ms4525Detect(pitotDev_t * pitot)
     uint8_t rxbuf[4];
     bool ack = false;
 
-    pitot->busDev = busDeviceInit(BUSTYPE_I2C, DEVHW_MS4525, 0, OWNER_AIRSPEED);
+    pitot->busDev = busDeviceInit(BUSTYPE_I2C, DEVHW_MS4525_0 + pitotmeterConfig()->ms4525_i2c_address, 0, OWNER_AIRSPEED);
```
Comment (JulioCesarMatias) [inline src/main/drivers/pitotmeter/pitotmeter_ms4525.c]: I just added the value configured in the parameter during sensor initialization.

Captura de tela 2021-07-13 132159

**Comment by JulioCesarMatias** [comment]
PR always returning error because of "settings.rb", I don't know how to solve it.

**Comment by stronnag** [comment]
Your variable name exceeds the somewhat undocumented restrictions in the 'name compressor' which saves us a huge amount of flash space. Please see https://github.com/iNavFlight/inav/pull/4090, which describes the limitations. We should really document this, as it's a real bitch to debug and discover the reason.

**Comment by stronnag** [comment]
To explain a bit more (no pun intended). We have an alphabet of a-z,NUL and 5 other characters (5 bits, 32 character set). This character set is shared across all names. 
Prior to this PR, we had added _, 1, 2,3 (i.e. 4 of the 5). 
You can have one extra (so either 4 or 5, not both).
Or we change the compressor to 6 bits, which will cost perhaps 500 bytes.

So just reducing the name length and keeping both 4 and 5 will not do.

**Comment by JulioCesarMatias** [comment]
> Your variable name exceeds the somewhat undocumented restrictions in the 'name compressor' which saves us a huge amount of flash space. Please see #4090, which describes the limitations. We should really document this, as it's a real bitch to debug and discover the reason.

it's strange, because there are other parameters with the same amount of bits, and they don't give an error. I was wondering if the numbers contained in the parameter name might be an "ms4525" problem.

**Comment by JulioCesarMatias** [comment]
> To explain a bit more (no pun intended). We have an alphabet of a-z,NUL and 5 other characters (5 bits, 32 character set). This character set is shared across all names.
> Prior to this PR, we had added _, 1, 2,3 (i.e. 4 of the 5).
> You can have one extra (so either 4 or 5, not both).
> Or we change the compressor to 6 bits, which will cost perhaps 500 bytes.
> 
> So just reducing the name length and keeping both 4 and 5 will not do.


removing the numbers from the parameter name, PR run normally. The problem was the numbers in the name, not the size.

**Comment by stronnag** [comment]
> > To explain a bit more (no pun intended). We have an alphabet of a-z,NUL and 5 other characters (5 bits, 32 character set). This character set is shared across all names.
> > Prior to this PR, we had added _, 1, 2,3 (i.e. 4 of the 5).
> > You can have one extra (so either 4 or 5, not both).
> > Or we change the compressor to 6 bits, which will cost perhaps 500 bytes.
> > So just reducing the name length and keeping both 4 and 5 will not do.
> 
> removing the numbers from the parameter name, PR run normally. The problem was the numbers in the name, not the size.
> But thanks for the comment, because he was the one who made me think about the numbers in the parameter name. Thanks!

It was exceeding the size of *character set*. You can have 1,2,3 and *one* other number.

File: src/main/fc/settings.yaml [hunk 2]
```diff
@@ -693,6 +693,11 @@ groups:
         min: 0
         max: 100
         default_value: 1.0
+      - name: airspeed_address
```
Comment (DzikuVx) [inline src/main/fc/settings.yaml:696]: let's rename it to `i2c_address_ms4525`

**Review by DzikuVx — CHANGES_REQUESTED** [review]

File: src/main/fc/settings.yaml (hunk 2 above)
Comment (stronnag) [reply src/main/fc/settings.yaml:696]: You can't do that without increasing the size of the compressed name character set, which will cost around 500b on every target.,

Comment (stronnag) [reply src/main/fc/settings.yaml:696]: * The character set is shared across all settings names
* It is currently 5 bit (32 characters)
* We use 31 with NUL, '_' , a-z, 1-3
* You can't have both 4 & 5

Comment (DzikuVx) [reply src/main/fc/settings.yaml:696]: lovely

Comment (stronnag) [reply src/main/fc/settings.yaml:696]: trivial to go to a 6 bit set if we're prepared to endure the size increase

**Review by stronnag — CHANGES_REQUESTED** [review]
Please also update the PG Template version (2 -> 3) as the structure has changed.

File: src/main/fc/settings.yaml (hunk 2 above)
Comment (JulioCesarMatias) [reply src/main/fc/settings.yaml:696]: I don't know what new name to put, because "ms4525" won't be possible

**Review by stronnag — APPROVED** [review]

File: src/main/fc/settings.yaml (hunk 2 above)
Comment (OptimusTi) [reply src/main/fc/settings.yaml:696]: Convert 4525 to hex 0x11AD?

Comment (JulioCesarMatias) [reply src/main/fc/settings.yaml:696]: > Convert 4525 to hex 0x11AD?


no, the lay user cannot read in hexadecimal.

Comment (stronnag) [reply src/main/fc/settings.yaml:696]: > Convert 4525 to hex 0x11AD?

In any case, the compression character set only contains lower case letters ...

**Comment by JulioCesarMatias** [comment]
when https://github.com/iNavFlight/inav/pull/7277 it's merged, I'll do a new push with the parameter renamed to `i2c_address_ms4525`

**Comment by JulioCesarMatias** [comment]
since migrating the table from 5 bits to 6 bits is feasible at the moment, so a workaround must be found. I thought about removing the config parameter from settings.yaml, and adding a function to do an automatic search for the pitot tube on the I2C bus.

example:

```
enum {
   MS4525_ADDR_0x28 = 0,
   MS4525_ADDR_0x36,
   MS4525_ADDR_0x46,
   MS4525_NOT_FOUND
};

bool ms4525Detect(pitotDev_t * pitot, uint8_t parse_i2c_addr_config)
{
    uint8_t rxbuf[4];
    uint8_t try_again_with_other_addr = MS4525_ADDR_0x28;
    bool ack = false;
    
    switch (try_again_with_other_addr) {

     case MS4525_ADDR_0x28: 
      pitot->busDev = busDeviceInit(BUSTYPE_I2C, DEVHW_MS4525_0 + try_again_with_other_addr, 0, OWNER_AIRSPEED);
     if (pitot->busDev == NULL) {
        try_again_with_other_addr = MS4525_ADDR_0x36
    }
   FALLTHROUGH;

    case MS4525_ADDR_0x36: 
      pitot->busDev = busDeviceInit(BUSTYPE_I2C, DEVHW_MS4525_0 + try_again_with_other_addr, 0, OWNER_AIRSPEED);
     if (pitot->busDev == NULL) {
        try_again_with_other_addr = MS4525_ADDR_0x46;
    }
   FALLTHROUGH;

    case MS4525_ADDR_0x46: 
      pitot->busDev = busDeviceInit(BUSTYPE_I2C, DEVHW_MS4525_0 + try_again_with_other_addr, 0, OWNER_AIRSPEED);
     if (pitot->busDev == NULL) {
         try_again_with_other_addr = MS4525_NOT_FOUND;
    }
   FALLTHROUGH;

    case MS4525_NOT_FOUND: 
     if (pitot->busDev == NULL) {
       return false;
     }
   FALLTHROUGH;
     
   }
```

**Comment by JulioCesarMatias** [comment]
> as 5-bit to 6-bit migration is not feasible at the moment, so a workaround must be found. I thought about removing the config parameter from settings.yaml, and adding a function to do an automatic search for the pitot tube on the I2C bus.
> 
> example:
> 
> ```
> enum {
>    MS4525_ADDR_0x28 = 0,
>    MS4525_ADDR_0x36,
>    MS4525_ADDR_0x46,
>    MS4525_NOT_FOUND
> };
> 
> bool ms4525Detect(pitotDev_t * pitot)
> {
>     uint8_t rxbuf[4];
>     uint8_t try_again_with_other_addr = MS4525_ADDR_0x28;
>     bool ack = false;
>     
>     switch (try_again_with_other_addr) {
> 
>      case MS4525_ADDR_0x28: 
>       pitot->busDev = busDeviceInit(BUSTYPE_I2C, DEVHW_MS4525_0 + try_again_with_other_addr, 0, OWNER_AIRSPEED);
>      if (pitot->busDev == NULL) {
>         try_again_with_other_addr = MS4525_ADDR_0x36
>     }
>    FALLTHROUGH;
> 
>     case MS4525_ADDR_0x36: 
>       pitot->busDev = busDeviceInit(BUSTYPE_I2C, DEVHW_MS4525_0 + try_again_with_other_addr, 0, OWNER_AIRSPEED);
>      if (pitot->busDev == NULL) {
>         try_again_with_other_addr = MS4525_ADDR_0x46;
>     }
>    FALLTHROUGH;
> 
>     case MS4525_ADDR_0x46: 
>       pitot->busDev = busDeviceInit(BUSTYPE_I2C, DEVHW_MS4525_0 + try_again_with_other_addr, 0, OWNER_AIRSPEED);
>      if (pitot->busDev == NULL) {
>          try_again_with_other_addr = MS4525_NOT_FOUND;
>     }
>    FALLTHROUGH;
> 
>     case MS4525_NOT_FOUND: 
>      if (pitot->busDev == NULL) {
>        return false;
>      }
>    FALLTHROUGH;
>      
>    }
> ```

i know this will make booting slower but i don't see any other alternative for the moment

**Pull request closed without merging**